### DISCUSSION TITLE
SearchParams for instant states

### DISCUSSION THREAD

**Discussion by AmirL** [vercel/next.js#56399] [App Router]
Hi,

It's evident that UI states often need to be reflected in the URL as well.
The issue is that useRouter and useSearchParams do not change state instantly,
unlike useState. (I've made [an example on codesandbox](https://codesandbox.io/p/sandbox/adoring-sun-64ppdw?file=%2Fapp%2Fpage.tsx%3A18%2C2) if you have doubts about this. Try to enable Network Throttling in your browser).

```ts
function QueryButton() {
  const router = useRouter();
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const params = new URLSearchParams(searchParams.toString());
  const state = searchParams.get("query") === "1";

  return (
    <div
      onClick={() => {
        params.set("query", state ? "0" : "1");
        router.push(pathname + "?" + params.toString(), { scroll: false });
      }}
      className={`${
        state ? "bg-red-500" : "bg-blue-500"
      } p-3 rounded-md text-white cursor-pointer`}
    >
      Query button
    </div>
  );
}
```

It appears that NextJs reloads the page when the URL changes, and sometimes it's unnecessary,
because the state has already changed.

How can I change the URL without requesting the page again? And still be able to use useSearchParams or any other reactive replacement for SearchParams?

**Comment by franky47** [comment]
I encountered this issue when building [`nuqs`](https://nuqs.47ng.com). The Next.js app router doesn't support shallow updates yet, so we have to go around it and use the Web History API directly.

Your example would then look like this:

```tsx
import { useQueryState, parseAsInteger } from 'nuqs'

function QueryButton() {
  const [state, setState] = useQueryState('query', parseAsInteger.withDefault(0))

  return (
    <div
      onClick={() => setState(x => 1 - x)} // toggle between 0 and 1
      className={`${
        state ? "bg-red-500" : "bg-blue-500"
      } p-3 rounded-md text-white cursor-pointer`}
    >
      Query button
    </div>
  );
}
```

**Reply by IGassmann** [reply]
Shallow updates are [supported by Next.js with the Web History API](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#using-the-native-history-api). [usePathname](https://nextjs.org/docs/app/api-reference/functions/use-pathname) and [useSearchParams](https://nextjs.org/docs/app/api-reference/functions/use-search-params) can be used alongside with it.

**Reply by franky47** [reply]
Yes, nuqs (the rebranding of `next-usequerystate`) supports both approaches through the `shallow` prop, much like the pages router did.

https://nuqs.47ng.com/docs/options#shallow

**Comment by AmirL** [comment]
Nice solution @franky47!

I ended with a similar approach:
```typescript
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import { useMemo, useState, useEffect } from 'react';

type toString = {
  toString(): string;
};

export type SetQueryParam<T extends toString> = (
  value: T,
  // Set it true on the last param update, and false when you need to update multiple params at once
  updateUrl?: boolean,
  // if prevParams is provided, it will use it instead of current url params
  // this is useful when you want to update multiple params at once
  prevParams?: URLSearchParams
) => URLSearchParams;

/**
 * Hook for replace useState with sync value with url.
 * It uses useState internally to make changes instantly,
 * but also syncs with url in case url is changed (back button clicked, or url changed manually)
 **/
export function useQueryParam<T extends toString>(key: string, validate: (value: string | null) => T) {
  const router = useRouter();
  const pathname = usePathname();
  const searchParams = useSearchParams();
  const queryValueRaw = searchParams.get(key);
  // validate query value
  const queryValue = useMemo(() => validate(queryValueRaw), [queryValueRaw, validate]);

  const [value, setValue] = useState<T>(queryValue);
  const valueStr = useMemo(() => value.toString(), [value]);

  const setQueryParam: SetQueryParam<T> = (newValue, updateUrl = true, prevParams?) => {
    const newValueStr = newValue.toString();
    const params = prevParams || new URLSearchParams(searchParams.toString());
    if (newValueStr !== queryValueRaw) {
      setValue(newValue);
      params.set(key, newValueStr);
      if (updateUrl) {
        router.push(pathname + '?' + params.toString(), { scroll: false });
      }
    }
    return params;
  };

  // sync value with url, in case url is changed (back button for example)
  useEffect(() => {
    if (queryValueRaw !== valueStr) {
      setValue(queryValue);
    }
    // Don't sync when state value is changed, it's a one way sync
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [queryValue, queryValueRaw]);

  return [value, setQueryParam] as const;
}
```

And an example of usage with Zod:
```typescript
export const SortDirectionSchema = z.enum(['asc', 'desc']).catch('asc');
export type SortDirection = z.infer<typeof SortDirectionSchema>;

//...

const [sortDirection, setSortDirection] = useQueryParam('sortDirection', (value) => SortDirectionSchema.parse(value));
```

**Comment by matheusgabardo** [comment]
Thanks guys, my solution for now was to add a state to mark when the url is updated 
<img width="441" alt="Captura de Tela 2024-06-17 às 17 35 36" src="https://github.com/vercel/next.js/assets/62657173/6b3e7d45-8e9f-4e33-9113-ec8c909ce674">

**Comment by IGassmann** [comment]
There are two solutions if you want to get instant states when modifying search params:

1. If your server-provided state doesn't depend on the search param, meaning you don't need to refetch data from the server after you've modified the search param, then you can [use the `window.history` API to update the search params.](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#using-the-native-history-api). This is officially supported by Next.js. There won't be any request to the server and the `useSearchParams()` and `usePathname()` hooks will update right away.

2. If your server-provided state depends on the search param (e.g., filtering upon server-provided data), then you can use `router.push()` along with `useOptimistic()`. `router.push()` will make sure to trigger a request to the server so that you can get fresh data, while `useOptimistic()` will update the UI instantaneously until the server respond. See this guide on how to implement this solution: https://buildui.com/posts/instant-search-params-with-react-server-components#solution-useoptimistic

**Comment by asmyshlyaev177** [comment]
Created a library for this - https://github.com/asmyshlyaev177/state-in-url  , can store complex state and enjoy TS autocomplete.
It use `window.history` for navigation in Next.js, no `_rsc` requests when updating searchParams.

```typescript
export const countState = { count: 0, other: { arr: [1,2,3 ] } }
...
import { useUrlState } from 'state-in-url/next';
import { countState } from './countState';

function MyComponent() {
  const { state, updateState, updateUrl } = useUrlState(countState);
...
      <button onClick={() => updateUrl({ count: state.count + 1 }) }>
     // or  <button onClick={(curr) => updateUrl({ ...curr, count: curr.count + 1 }) }>
        Increment
      </button>
```

**Comment by pplytas** [comment]
In my case using **Next.js v15.0.2**, I achieved the optimistic URL updates with the following custom hook:

```typescript
import { useCallback } from "react";
import { useRouter, useSearchParams as next_useSearchParams } from "next/navigation";

export function useSearchParams() {
  const router = useRouter();
  const searchParams = next_useSearchParams();

  const setSearchParams = useCallback(
    (
      updateFn: (previousSearchParams: URLSearchParams) => URLSearchParams,
      options?: { replace?: boolean }
    ) => {
      const newSearchParams = updateFn(searchParams);

      // This updates the URL state immediately
      if (options?.replace) {
        window.history.replaceState({}, "", `?${newSearchParams.toString()}`);
      } else {
        window.history.pushState({}, "", `?${newSearchParams.toString()}`);
      }

      // Schedule the router refresh for the next tick to get the server response as well
      setTimeout(() => {
        router.refresh();
      }, 0);
    },
    [router, searchParams]
  );

  return [searchParams, setSearchParams];
}
```

and then using it as such:

```typescript
'use client';

function Component() {
  const [searchParams, setSearchParams] = useSearchParams();

  console.log(searchParams); // Logs the updated search params immediately and a second time when the server responds

  return (
    <button
      onClick={() => {
        setSearchParams((prevSearchParams) => {
          const newSearchParams = new URLSearchParams(prevSearchParams.toString());
          // Do something with newSearchParams
          return newSearchParams;
        });
      }}
    >
      Click me
    </button>
  );
}
```

**Comment by asmyshlyaev177** [comment]
Next.js doesn't change state instantly because it requrests server components even when you change searchParams, you can see `_rsc` requests, there is a discussion https://github.com/vercel/next.js/discussions/59167 .
Can change the url with `window.history` pushState/replaceState, it will happen instantly.
